libraries/persistent_store/src/buffer.rs

```diff
@@ -0,0 +1,668 @@
+    ///
+    /// [armed]: struct.BufferStorage.html#method.arm_interruption
+    /// [corruption function]: type.BufferCorruptFunction.html
+    pub fn corrupt_operation(&mut self, corrupt: BufferCorruptFunction) {
+        let operation = self.interruption.get().unwrap();
+        let range = self.operation_range(&operation).unwrap();
+        let mut before = self.storage[range.clone()].to_vec().into_boxed_slice();
+        match operation {
+            BufferOperation::Write { value: after, .. } => {
+                corrupt(&mut before, &after);
+                self.incr_word_writes(range.start, &before, &after);
+            }
+            BufferOperation::Erase { page } => {
+                let after = vec![0xff; self.page_size()].into_boxed_slice();
+                corrupt(&mut before, &after);
+                if before == after {
+                    self.incr_page_erases(page);
+                }
+            }
+        };
+        self.storage[range].copy_from_slice(&before);
+    }
+
+    /// Returns the number of times a word was written.
+    pub fn get_word_writes(&self, word: usize) -> usize {
+        self.word_writes[word]
+    }
+
+    /// Returns the number of times a page was erased.
+    pub fn get_page_erases(&self, page: usize) -> usize {
+        self.page_erases[page]
+    }
+
+    /// Sets the number of times a page was erased.
+    pub fn set_page_erases(&mut self, page: usize, cycle: usize) {
+        self.page_erases[page] = cycle;
+    }
+
+    /// Returns whether a number is word-aligned.
+    fn is_word_aligned(&self, x: usize) -> bool {
+        x & (self.options.word_size - 1) == 0
+    }
+
+    /// Returns whether a number is page-aligned.
+    fn is_page_aligned(&self, x: usize) -> bool {
+        x & (self.options.page_size - 1) == 0
+    }
+
+    /// Updates the counters as if a page was erased.
+    ///
+    /// The page counter of that page is incremented and the word counters of that page are reset.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the maximum number of erase cycles per page is reached.
+    fn incr_page_erases(&mut self, page: usize) {
+        assert!(self.page_erases[page] < self.max_page_erases());
+        self.page_erases[page] += 1;
+        let num_words = self.page_size() / self.word_size();
+        for word in 0..num_words {
+            self.word_writes[page * num_words + word] = 0;
+        }
+    }
+
+    /// Updates the word counters as if a partial write occurred.
+    ///
+    /// The partial write is described as if `complete` was supposed to be written to the storage
+    /// starting at byte `index`, but actually only `value` was written. Word counters are
+    /// incremented only if their value would change and they would be completely written.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the maximum number of writes per word is reached.
+    fn incr_word_writes(&mut self, index: usize, value: &[u8], complete: &[u8]) {
```

Comment:
Yes, this is implicit in the fact "as if a partial write occurred" since writes are word-aligned. Should I make it explicit? (The same applies for the length of the slices which should also be of the same length.)

Reply:
You have been doing that usually. So I think mentioning that the caller is responsible makes sense.

Reply:
Okay, your choice. Feel free to merge or I can re-approve if you want to change the documentation.

Reply:
I've already changed the documentation in the previous commit.